src/v/cluster/partition_manager.cc

```diff
@@ -39,18 +39,46 @@ ss::future<consensus_ptr> partition_manager::manage(
   storage::ntp_config ntp_cfg,
   raft::group_id group,
   std::vector<model::broker> initial_nodes) {
-    return _storage.log_mgr()
-      .manage(std::move(ntp_cfg))
-      .then([this, group, nodes = std::move(initial_nodes)](
-              storage::log&& log) mutable {
-          return _raft_manager.local()
-            .create_group(group, std::move(nodes), log)
-            .then([this, log, group](ss::lw_shared_ptr<raft::consensus> c) {
-                auto p = ss::make_lw_shared<partition>(c, _tx_gateway_frontend);
-                _ntp_table.emplace(log.config().ntp(), p);
-                _raft_table.emplace(group, p);
-                _manage_watchers.notify(p->ntp(), p);
-                return p->start().then([c] { return c; });
+    return ss::do_with(
+      std::move(ntp_cfg),
+      [this, group, initial_nodes = std::move(initial_nodes)](
+        storage::ntp_config& ntp_cfg) mutable {
+          return _storage.downloader().download_log(ntp_cfg).then(
```

Comment:
we can only download the data when 
- initial nodes are assigned to the partition replica
- the initial topic revision is known (that means that command should already be submitted to the controller log)

The startup is not affected since the code is invoked only when topic is actually created, the only thing that is done every time on startup is the check that the directory exist. The check is needed to skip download during future log replays. It could be replaced with something else. And the filesystem based check is fast (5-10 microseconds).

Reply:
also, the recovery happens rarely, during disaster recovery or during recovery after accidental deletion. Based on this, I'm not sure that we should throttle this process. The actual recovery can be throttled using scheduling group of cause but it's not very different from normal archival workload to begin with.